test/absinthe/introspection_test.exs

```diff
@@ -89,6 +89,88 @@ defmodule Absinthe.IntrospectionTest do
     end
   end
 
+  describe "introspection of complex directives" do
+    defmodule ComplexDirectiveSchema do
+      use Absinthe.Schema
+      use Absinthe.Fixture
+
+      defmodule ComplexDirectivePrototype do
+        use Absinthe.Schema.Prototype
+
```

Comment:
@jeffutter thank you so much for looking into this!

Would you be able to add a scalar and an enum type to this prototype and use those in the directive arguments to ensure that's handled as well?

Reply:
@jeffutter I can't resolve my previous comment for some reason so I'll respond to your last message here

> I just realized that the fixed test is broken with `SCHEMA_PROVIDER=persistent_term` (didn't realize that setting was run in CI). So I'll implement `persistent_term` shortly and expand the test as @kzlsakal suggested.
> 
> Done: ✅

I did some testing with our service which uses both absinthe_federation and absinthe. I think this change is super close (thanks for adding the fixes for persistent term as well). Although during my testing, I found that absinthe_federation specifies a `name` for the `:_field_set` scalar [here](https://github.com/DivvyPayHQ/absinthe_federation/blob/main/lib/absinthe/federation/schema/prototype/federated_directives.ex#L12) which starts with an underscore. When I tried to change apply this to your test and changed the scalar definition to 

```elixir
scalar :_normal_string, name: "_NormalString", description: "string" do
```

The output of the `complex_directive_args` you're testing below became this:

```elixir
[
  %{"defaultValue" => nil, "description" => nil, "name" => "colorChannel", "type" => %{"kind" => "ENUM", "name" => "ColorChannel"}},
  %{"defaultValue" => nil, "description" => nil, "name" => "complex", "type" => %{"kind" => "INPUT_OBJECT", "name" => "Complex"}},
  %{"defaultValue" => nil, "description" => nil, "name" => "normalString", "type" => %{"kind" => "SCALAR", "name" => "_NormalString"}}
]
```

as you can see while the "type" is correct, the underscore from the name of its parent is stripped. This eventually causes the following error when I try to run the introspection query on the actual endpoint that uses absinthe federation:

```
Error: Invalid or incomplete schema, unknown type: _FieldSet. Ensure that a full introspection query is used in order to build a client schema.
```

hope that helps!

Reply:
Although I can confirm this PR resolves the issue with the introspection query receiving a GraphQL error (`Cannot return null for non-nullable field`) in the first place.

Reply:
I think the introspection parser expects those types to be defined in the __schema.types path when you add the following to your query:

```graphql
query IntrospectionQuery {
  __schema {
    types {
      name
    }
    directives {
      ...
```

Reply:
Hmmm. I was using a very simple project with AbsintheFederation and was able to print the introspection query. Perhaps it was too simple to hit that. I’ll try to construct a more complex example over the weekend.

Reply:
@jeffutter just tested after your [1f8122e](https://github.com/absinthe-graphql/absinthe/pull/1334/commits/1f8122edd524dd4850e95a548bc62077d08b42a4) update. Works like a charm 👏

Reply:
Woo that was fast :-D 💨 . I just verified with GraphiQL and came here to ping you, you beat me too it 🤣